src/physics/point_source.rs

```diff
@@ -0,0 +1,205 @@
+
+    // Assemble components
+    let c = 3.0 * rinv3;
+    let term1 = (rmr.0 * c, rmr.1 * c, rmr.2 * c);
+    let term2 = (-moment.0 * rinv3, -moment.1 * rinv3, -moment.2 * rinv3);
+    let tsum = (term1.0 + term2.0, term1.1 + term2.1, term1.2 + term2.2);
+
+    let (bx, by, bz) = (
+        MU0_OVER_4PI * tsum.0,
+        MU0_OVER_4PI * tsum.1,
+        MU0_OVER_4PI * tsum.2,
+    );
+
+    (bx, by, bz) // [T]
+}
+
+/// Magnetic flux density of a dipole in cartesian coordiantes.
+///
+/// Arguments
+///
+/// * loc: (m) location of the point source
+/// * moment: (A-m^2) magnetic moment vector of the point source
+/// * obs: (m) observation point to examine
+/// * out: (T) storage for B-field
+///
+/// Returns
+///
+/// * (bx, by, bz) [T] magnetic field components at observation point
+#[inline]
+pub fn flux_density_dipole(
+    loc: (&[f64], &[f64], &[f64]),
+    moment: (&[f64], &[f64], &[f64]),
+    obs: (&[f64], &[f64], &[f64]),
+    out: (&mut [f64], &mut [f64], &mut [f64]),
+) -> Result<(), &'static str> {
+    // Check lengths
+    let m = loc.0.len();
+    let n = obs.0.len();
+
+    check_length_3tup!(m, &loc);
+    check_length_3tup!(m, &moment);
+    check_length_3tup!(n, &obs);
+    check_length_3tup!(n, &out);
+
+    // Do calcs
+    for i in 0..n {
+        for j in 0..m {
+            let obsi = (obs.0[i], obs.1[i], obs.2[i]);
+            let locj = (loc.0[j], loc.1[j], loc.2[j]);
+            let momentj = (moment.0[j], moment.1[j], moment.2[j]);
+            let (bx, by, bz) = flux_density_dipole_scalar(locj, momentj, obsi);
+            out.0[i] += bx;
+            out.1[i] += by;
+            out.2[i] += bz;
+        }
+    }
+
+    Ok(())
+}
+
+/// Magnetic flux density of a dipole in cartesian coordiantes.
+/// Parallelized over chunks of observation points and vectorized over source points.
+///
+/// Arguments
+///
+/// * loc: (m) location of the point source
+/// * moment: (A-m^2) magnetic moment vector of the point source
+/// * obs: (m) observation point to examine
+/// * out: (T) storage for B-field
+///
+/// Returns
+///
+/// * (bx, by, bz) [T] magnetic field components at observation point
+#[inline]
+pub fn flux_density_dipole_par(
+    loc: (&[f64], &[f64], &[f64]),
+    moment: (&[f64], &[f64], &[f64]),
+    obs: (&[f64], &[f64], &[f64]),
+    out: (&mut [f64], &mut [f64], &mut [f64]),
+) -> Result<(), &'static str> {
+    // Chunk inputs
+    let n = chunksize(obs.0.len());
+    let (obsxc, obsyc, obszc) = par_chunks_3tup!(obs, n);
+    let (outxc, outyc, outzc) = mut_par_chunks_3tup!(out, n);
+
+    outxc
+        .zip(outyc.zip(outzc.zip(obsxc.zip(obsyc.zip(obszc)))))
+        .try_for_each(|(outx, (outy, (outz, (obsx, (obsy, obsz)))))| {
+            flux_density_dipole(loc, moment, (obsx, obsy, obsz), (outx, outy, outz))
+        })?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use std::f64::consts::PI;
+
+    use crate::testing::*;
+
+    // Make sure that far from a dipole on the Z axis, the field
```

Comment:
Nitpick: The test does not just check observation points on the z axis, but over a unit cube. That's good, but this comment should reflect that.
```suggestion
    // Make sure that far from a dipole, the field
```